content/news/2024-02-03-bevy-0.13/index.md

```diff
@@ -0,0 +1,1518 @@
+  * Events can be cleared before a stepped system can read them
+  * Best approach here is to configure event-based systems to always run,
+    or put them in a schedule not added to `Stepping`
+  * Continue with breakpoints may also work in this scenario
+* Conditional systems may not run as expected when stepping
+  * Similar to event-based systems, if the condition is true for only a short
+    time, system may not run when stepped
+
+### Detailed Examples
+
+* [Text-based stepping example](https://github.com/bevyengine/bevy/blob/main/examples/ecs/system_stepping.rs)
+* Non-interactive [bevy UI example stepping plugin](https://github.com/bevyengine/bevy/blob/main/examples/games/stepping.rs) used in the breakout example
+* Interactive [egui stepping plugin](https://gist.github.com/dmlary/3fd57ebf1f88bb9afa8a6604737dac97) used in demo videos
+
+[`Stepping`]: https://docs.rs/bevy/0.13.0/bevy/ecs/schedule/stepping/Stepping.html
+
+## Query transmutation
+
+<div class="release-feature-authors">authors: @hymm, james-j-obrien</div>
+
+Have you every wanted to pass a query to a function, but instead of having a
+`Query<&Transform>` you have a `Query<(&Transform, &Velocity), With<Enemy>>`?
+Well now you can by using the `Query::transmute_lens` method. Query transmutes
+allow you to change a query into different query types as long as the
+components accessed are a subset of the original query. If you do try to access
+data that is not in the original query, this method will panic.
+
+```rust
+fn reusable_function(lens: &mut QueryLens<&Transform>) {
+    let query = lens.query();
+    // do something with the query...
+}
+
+// We can use the function in a system that takes the exact query.
+fn system_1(mut query: Query<&Transform>) {
+    reusable_function(&mut query.as_query_lens());
+}
+
+// We can also use it with a query that does not match exactly
+// by transmuting it.
+fn system_2(mut query: Query<(&mut Transform, &Velocity), With<Enemy>>) {
+    let mut lens = query.transmute_lens::<&Transform>();
+    reusable_function(&mut lens);
+}
+```
+
+Note that the `QueryLens` will still iterate over the same entities as the
+original `Query` it is derived from. A `QueryLens<&Transform>` taken from
+a `Query<(&Transform, &Velocity)>`, will only include the `Transform` of
+entities with both `Transform` and `Velocity` components.
+
+Besides removing parameters you can also change them in limited ways to the
+different smart pointer types. One of the more useful is to change a
+`& mut` to a `&`. See the [documentation](https://docs.rs/bevy/latest/bevy/ecs/system/struct.Query.html#method.transmute_lens)
+for more details.
+
+One thing to take into consideration is the transmutation is not free.
+It works by creating a new state and copying a bunch of the cached data
+inside the original query. It's not a expensive operation, but you should
+probably avoid doing it inside a hot loop.
+
+## Entity Optimizations
```

Comment:
This section is extremely technical and goes into way more detail than feels appropriate for a release blog post. It's also way too early in the blog. IMO the sections should be ordered as to highlight the most exciting user-facing high-level features first, and the most technical low-level improvements to engine internals last.

I suggest to either simplify / shorten it, omitting the most low-level details.

Also please move it way further down towards the end of the blog.

Reply:
First off, I want to add my agreement that this probably should go more towards the end of the blog, for the exact reasons above: User facing stuff first, technical/perf improvements last.

Simplifying this will probably entail removing the details/context as to _why_ such a change was important (but also why it didn't land previously). This sort of exploration is I think beneficial to the rust community at large that could benefit from the visibility of the techniques used/merged into bevy. *That said*, maybe this is a candidate for a follow-up post? A post where we can indeed go into more depth about these topics. Summarise for the release, expand in the follow-up.

Likewise, we could take the opportunity to write more of that sort of content to the bevy website. There's no shortage of stuff we could talk about in terms of highly technical challenges Bevy faces in trying to optimise, improve and further the state of art for ECS in Rust and in the game engine space. Also a nice way to keep some visibility/activity for the Bevy project outside of Discord instead of every 3/4 months (apart from the occasional other post).

Reply:
Yeah, I really like this content, but I've been chewing on how to reduce its length. I'm personally in favor of focusing on impact, and moving the crunchy details into either an official blog post, or a third-party blog post that we boost.

Reply:
This has been greatly reduced and moved to the end <3 Super cool stuff, but distracting.